File: internal/application/service/ingredient_loader_service.go
package service

import (
	"github.com/vini/costify-go/internal/application/contract"
	"github.com/vini/costify-go/internal/application/errors"
	"github.com/vini/costify-go/internal/domain/entity"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// IngredientLoaderService loads ingredients for recipe processing
type IngredientLoaderService struct {
	ingredientRepository contract.IngredientRepository
}

// NewIngredientLoaderService creates a new IngredientLoaderService
func NewIngredientLoaderService(ingredientRepository contract.IngredientRepository) *IngredientLoaderService {
	return &IngredientLoaderService{
		ingredientRepository: ingredientRepository,
	}
}

// LoadIngredients loads ingredients from recipe ingredients list
func (s *IngredientLoaderService) LoadIngredients(
	recipeIngredients []valueobject.RecipeIngredient,
) (map[valueobject.Id]*entity.Ingredient, error) {
	ingredientMap := make(map[valueobject.Id]*entity.Ingredient)

	for _, recipeIngredient := range recipeIngredients {
		ingredientId := recipeIngredient.IngredientId()

		ingredient, err := s.ingredientRepository.FindById(ingredientId)
		if err != nil {
			return nil, errors.NewIngredientNotFoundError("Ingredient not found with ID: " + ingredientId.Value())
		}

		ingredientMap[ingredientId] = ingredient
	}

	return ingredientMap, nil
}

File: api/src/main/java/br/unifor/costify/application/service/IngredientLoaderService.java
package br.unifor.costify.application.service;

import br.unifor.costify.application.contracts.IngredientRepository;
import br.unifor.costify.application.errors.IngredientNotFoundException;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.RecipeIngredient;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.springframework.stereotype.Service;

@Service
public class IngredientLoaderService {
  private final IngredientRepository ingredientRepository;

  public IngredientLoaderService(IngredientRepository ingredientRepository) {
    this.ingredientRepository = ingredientRepository;
  }

  public Map<Id, Ingredient> loadIngredients(List<RecipeIngredient> recipeIngredients) {
    Map<Id, Ingredient> ingredientMap = new HashMap<>();
    
    for (RecipeIngredient recipeIngredient : recipeIngredients) {
      Id ingredientId = recipeIngredient.getIngredientId();
      
      Optional<Ingredient> ingredientOpt = ingredientRepository.findById(ingredientId);
      if (ingredientOpt.isEmpty()) {
        throw new IngredientNotFoundException("Ingredient not found with ID: " + ingredientId);
      }
      
      ingredientMap.put(ingredientId, ingredientOpt.get());
    }
    
    return ingredientMap;
  }
}